### PR TITLE
Megatron GPT model finetuning

### PR DESCRIPTION
**Pull request by MaximumEntropy**
# What does this PR do ?

Adds the ability to fine-tune Megatron GPT Models.

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  answer_only_loss: False # not used right now
+  gradient_as_bucket_view: False
+
+  hidden_dropout: 0.0
+  attention_dropout: 0.0
+  ffn_dropout: 0.0
+
+  data:
+    train_ds:
+      file_names: ??? # Path to a list of JSONL files corresponding to the source data.
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:70]: I'm a little confused what this should be. Could you give a commented example to make it clear?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      add_bos: ${model.data.train_ds.add_bos}
+      separate_prompt_and_response_with_newline: ${model.data.train_ds.separate_prompt_and_response_with_newline}
+      write_predictions_to_file: False
+      output_file_path_prefix: null # Prefix of the file to write predictions to.
+      truncation_field: "context" # Options: ['context', 'answer']
+      index_mapping_dir: null # Path to a directory to write index mapping files.
+      assistant_prompt: null # Options: ['long', 'short', 'long_nv', null]
+
+      metric:
+        name: "exact_string_match" # Name of the evaluation metric to use.
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:114]: What are the possible choices?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      global_batch_size: ${model.global_batch_size}
+      micro_batch_size: ${model.micro_batch_size}
+      shuffle: True
+      num_workers: 4
+      pin_memory: True
+      max_seq_length: 2048
+      min_seq_length: 1
+      drop_last: True
+      concat_sampling_technique: temperature # When providing a list of datasets, this arg defines the sampling strategy. Options: ['temperature', 'random']
+      concat_sampling_temperature: 5 # When providing a list of datasets, this arg defines the sampling temperature when strategy='temperature'
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:130]: strategy is sampling_technique here?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+      index_mapping_dir: null # Path to a directory to write index mapping files.
+      assistant_prompt: null # Options: ['long', 'short', 'long_nv', null]
+
+      metric:
+        name: "exact_string_match" # Name of the evaluation metric to use.
+        average: null # Average the metric over the dataset. Options: ['macro', 'micro']. Works only for 'F1', 'accuracy' etc. Refer to torchmetrics for metrics where this is supported.
+        num_classes: null
+
+  optim:
+    name: fused_adam
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:150]: does distributed optimizer work for sft?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import tempfile
+
+import torch.multiprocessing as mp
```
[the remark this answers is not in the document]
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:18]: Have you run into any cuda init errors? If not using spawn then can remove this line

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+)
+from nemo.core.config import hydra_runner
+from nemo.utils import AppState, logging
+from nemo.utils.exp_manager import StatelessTimer, exp_manager
+from nemo.utils.model_utils import inject_model_parallel_rank
+
+
+def _modify_config(gpt_cfg, cfg, add_cfg_to_tree=False):
+    """
+    This function modifies the original gpt pre-training config (t5_cfg) with attributes from the finetuning config (cfg).
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:42]: t5_cfg?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    NLPSaveRestoreConnector,
+    PipelineMixedPrecisionPlugin,
+)
+from nemo.core.config import hydra_runner
+from nemo.utils import AppState, logging
+from nemo.utils.exp_manager import StatelessTimer, exp_manager
+from nemo.utils.model_utils import inject_model_parallel_rank
+
+
+def _modify_config(gpt_cfg, cfg, add_cfg_to_tree=False):
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:40]: Can we make this a general utility function? We are doing this modify config for many different scripts.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        # This is needed when modifying a hparam file directly to load `.ckpt` files.
+        # This is not needed to modify the cfg in `.nemo` files.
+        if add_cfg_to_tree:
+            OmegaConf.resolve(gpt_cfg)
+            gpt_cfg.cfg = gpt_cfg
+
+    return gpt_cfg
+
+
+def load_from_nemo(cls, cfg, trainer, gpt_cfg, modify_confg_fn):
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:76]: Same here can we make this a general utility function or classmethod?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    model = cls.restore_from(
+        restore_path=cfg.model.restore_from_path,
+        trainer=trainer,
+        override_config_path=gpt_cfg,
+        save_restore_connector=save_restore_connector,
+    )
+    return model
+
+
+def load_from_checkpoint_dir(cls, cfg, trainer, modify_confg_fn):
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:90]: Can we abstract this?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        resume_from_checkpoint = cfg.model.resume_from_checkpoint
+    else:
+        resume_from_checkpoint = trainer._checkpoint_connector.resume_from_checkpoint_fit_path
+    logging.info(f'Resuming training from checkpoint: {resume_from_checkpoint}')
+
+    trainer._checkpoint_connector = CheckpointConnector(trainer, resume_from_checkpoint=resume_from_checkpoint)
+    # Override timer callback to a stateless one
+    for idx, callback in enumerate(trainer.callbacks):
+        if isinstance(callback, Timer):
+            trainer.callbacks[idx] = StatelessTimer(cfg.trainer.max_time,)
```
Comment (ericharper) [inline examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:174]: We could abstract his as well

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 3 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:130]: This should disappear - I don't think we have any other strategy besides providing all sampling probs explicitly.

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 4 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:150]: It does, I can move the default to distributed_fused_adam

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                trainer=trainer,
+                return_config=True,
+                save_restore_connector=save_restore_connector,
+            )
+            model = load_from_nemo(MegatronGPTSFTModel, cfg, trainer, gpt_cfg, modify_confg_fn=_modify_config)
+        else:
+            validate_checkpoint_loading_args(cfg.model.pretrained_checkpoint)
+            model = load_from_checkpoint_dir(
+                MegatronGPTSFTModel, cfg, trainer, gpt_cfg, modify_confg_fn=_modify_config
+            )
```
[the remark this answers is not in the document]
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:192]: Too many args?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+                restore_path=cfg.model.restore_from_path,
+                trainer=trainer,
+                return_config=True,
+                save_restore_connector=save_restore_connector,
+            )
+            model = load_from_nemo(MegatronGPTSFTModel, cfg, trainer, gpt_cfg, modify_confg_fn=_modify_config)
+        else:
+            validate_checkpoint_loading_args(cfg.model.pretrained_checkpoint)
+            model = load_from_checkpoint_dir(
+                MegatronGPTSFTModel, cfg, trainer, gpt_cfg, modify_confg_fn=_modify_config
```
[the remark this answers is not in the document]
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:191]: Where does gpt_cfg come from in this case?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+            )
+    trainer.fit(model)
```
[the remark this answers is not in the document]
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:193]: what if train_ds doesn't have file_names?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+        label_key: str = "answer",
+        separate_prompt_and_response_with_newline: bool = False,
+        answer_only_loss: bool = True,
+        truncation_field: str = "answer",
+        pad_to_max_length: bool = True,
+        index_mapping_dir: str = None,
+        assistant_prompt: str = None,
+    ):
+        """
+        file_path: Path to a JSONL GPT supervised fine-tuning dataset.
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:49]: Would be nice if we can link to more information about what this is in the future

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+        assert self.truncation_field in ["answer", "context"]
+
+        self.indexed_dataset = JSONLMemMapDataset(dataset_paths=[file_path], tokenizer=None, header_lines=0)
+
+        # Will be None after this call if `max_num_samples` is None
+        self._build_samples_mapping()
+
+    def _build_samples_mapping(self):
+        if self.max_num_samples is not None:
+            # TODO: double-check this for decoder-only GPT
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:96]: still need this TODO?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+        tokenizer: Tokenizer for the dataset. Instance of a class that inherits TokenizerSpec (ex: YTTM, SentencePiece).
+        max_seq_length (int): maximum sequence length for each dataset examples. Examples will either be truncated to fit this length or dropped if they cannot be truncated.
+        min_seq_length (int): min length of each data example in the dataset. Data examples will be dropped if they do not meet the min length requirements. 
+        add_bos (bool): Whether to add a beginning of sentence token to each data example
+        add_eos (bool): Whether to add an end of sentence token to each data example
+        add_sep (bool): Whether to add a separation token to each data example (goes between prompt and answer)
+        tokens_to_generate (int): (inference only) Number of tokens to generate during inference
+        seed: Random seed for data shuffling.
+        max_num_samples: Maximum number of samples to load. This can be > dataset length if you want to oversample data. If None, all samples will be loaded.
+        seed: int = 1234,
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:59]: Would it be better to default this to None and get a random random seed every time?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+            self.samples_mapping = None
+
+    def __len__(self):
+        if self.max_num_samples is None:
+            return len(self.indexed_dataset)
+        else:
+            return len(self.samples_mapping)
+
+    def __getitem__(self, idx):
+        if isinstance(idx, np.int64):
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:119]: can we do a more general check if it's a numpy array?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+            idx = idx.item()
+
+        if self.samples_mapping is not None:
+            assert idx < len(self.samples_mapping)
+            idx, _, _ = self.samples_mapping[idx]
+            if isinstance(idx, np.uint32):
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:125]: same question here

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+        input_ids = input_ids + answer_ids
+
+        if self.add_bos:
+            input_ids = [self.tokenizer.bos_id] + input_ids
+            answer_start_idx += 1
+        if self.add_eos:
+            input_ids = input_ids + [self.tokenizer.eos_id]
+
+        # We return None if the example fails in the min_seq_length and max_seq_length checks
+        # TODO: verify on the dataloader how to skip Nones and keep constant batch size
```
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:191]: Still need this TODO?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 20]
```diff
@@ -0,0 +1,665 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:1]: 2023

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 21]
```diff
@@ -0,0 +1,665 @@
...
+                if 'rouge' not in data_cfg.metric.name:
+                    metric = [metric(average=data_cfg.metric.average, num_classes=data_cfg.metric.num_classes)]
+                else:
+                    metric = [metric()]
+
+        return metric, metric_name
+
+    @property
+    def _metrics_require_string2category_map(self):
+        return set(["f1", "accuracy", "average_precision"])
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:130]: why set here?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 22]
```diff
@@ -0,0 +1,665 @@
...
+                    )
+                )
+
+            if len(data_cfg.get('concat_sampling_probabilities', None)) != len(data_cfg.file_names):
+                raise ValueError(
+                    (
+                        f"concat_sampling_probabilities must be of the same size as file_names.",
+                        f"Provided size {len(data_cfg.concat_sampling_probabilities)}, number of datasets {len(data_cfg.file_names)}",
+                    )
+                )
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:195]: Maybe can move this to megatron base model? I think gpt pretraining could use the same checks?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 23]
```diff
@@ -0,0 +1,665 @@
...
+        # Call parent validation step to get the loss.
+        loss = super().validation_step(dataloader_iter, batch_idx)
+        # NOTE: This is a hack to to sidestep inference when the user wants to just monitor the loss only.
+        # TODO (sandeepsub): This should be a better check.
+        return {
+            'loss': loss,
+            'preds': None,
+            'labels': None,
+            'inputs': None,
+        }
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:280]: Don't you always return here?

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 4 above)
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:150]: Maybe just mention it in a comment

**Review by ericharper — CHANGES_REQUESTED** [review]
Excellent PR!

We can merge this without addressing my "refactoring" comments, but please see the other small comments. Also, it looks like code scanning found some real issues so please take a look at those as well.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 16 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:59]: I think we force a seed with pre-training which is useful for debugging.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 15 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:96]: Yep, just removed it.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 23 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:280]: Yeah I need to update the comment and leave a TODO to figure out generation in a subsequent PR.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 21 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:130]: Just to check membership faster. But yeah probably overkill given that its small.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 17 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:119]: This is actually just a single numpy integer and not an array. I guess I could check if its `np.int` which maybe checks for all np integer types?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 22 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:195]: GPT Pre-training data is provided differently where sampling probs and datasets are provided in a single argument. I wanted to separate it out, so it wouldn't translate directly.

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+      file_names: ??? # Path to a list of JSONL files corresponding to the source data.
+      global_batch_size: ${model.global_batch_size}
+      micro_batch_size: ${model.micro_batch_size}
+      shuffle: True
+      num_workers: 4
+      pin_memory: True
+      max_seq_length: 2048
+      min_seq_length: 1
+      drop_last: True
+      concat_sampling_probabilities: null # When providing a list of datasets, this arg defines the sampling probabilities from each dataset when strategy='random'
```
[the remark this answers is not in the document]
Comment (okuchaiev) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:79]: we are discussing whether we should have and which one recipe public.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:191]: Fixed. That was a mistake in argument passing.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:193]: Ah yes that was an artifact of incorrect copy pasta from T5 fine-tuning. Fixed.

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:18]: Done!

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 24 above)
[the remark this answers is not in the document]
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:79]: Actually, this is a mistake - the only that is allowed right now is to hand-specify each probability by hand. I've removed those args.

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 2 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:114]: Hah thats a good catch, Although I have string match here, the only that works right now is the loss.

File: examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml (hunk 4 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml:147]: Done!

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 6 above)
Comment (MaximumEntropy) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:42]: Hah copy-pasta errors.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 14 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:49]: I did add a little bit of info.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 20 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py]: Done!

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 19 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py]: Done!

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 9 above)
Comment (yidong72) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:89]: Or we can put this into a utility function. It is used a lot in other places to load from checkpoint dir

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+from nemo.collections.nlp.parts.nlp_overrides import (
+    GradScaler,
+    MegatronHalfPrecisionPlugin,
+    NLPDDPStrategy,
+    NLPSaveRestoreConnector,
+    PipelineMixedPrecisionPlugin,
+)
+from nemo.core.config import hydra_runner
+from nemo.utils import AppState, logging
+from nemo.utils.exp_manager import StatelessTimer, exp_manager
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:35]: remove unused imports?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+            # Make sure that '{output}' always occurs at the end of the prompt template string
+            assert self.prompt_template.index('{output}') == len(self.prompt_template) - len('{output}')
+            # Get the context by replacing only the input
+            original_context = context
+            context = self.prompt_template.replace('{input}', context).replace('{output}', '').strip(' ')
+            # Replace the input and output placeholders with the actual input and output
+            text = self.prompt_template.replace('{input}', original_context).replace('{output}', output)
+
+        if self.separate_prompt_and_response_with_newline and self.prompt_template is None:
+            text = context + '\n' + output
```
Comment (yidong72) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:155]: should we use user provided separators?

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+        Truncation is carried out when needed, but it is performed only on the prompt side.
+        BOS, EOS, and SEP, are added if specified.
+        """
+        context = example[self.context_key]
+        output = example[self.label_key]
+
+        if self.prompt_template is not None:
+            import ipdb
+
+            ipdb.set_trace()
```
Comment (yidong72) [inline nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:143]: remove the debug statement?

File: scripts/nlp_language_modeling/niv2/preprocess_niv2.py [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+import re
+from argparse import ArgumentParser
+from multiprocessing import Pool
+
+from sacremoses import MosesDetokenizer
```
Comment (yidong72) [inline scripts/nlp_language_modeling/niv2/preprocess_niv2.py:21]: is it part of the plan to release NIV and T0 data preprocessing scripts? We would like others to SFT GPT with the same instruction dataset?

**Review by yidong72 — COMMENTED** [review]
Loos good. left some comments.

File: nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py (hunk 26 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:155]: I think the prompt_template should cover this case right?

File: examples/nlp/language_modeling/tuning/megatron_gpt_sft.py (hunk 9 above)
Comment (ericharper) [reply examples/nlp/language_modeling/tuning/megatron_gpt_sft.py:88]: Yes exactly

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py (hunk 22 above)
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:195]: Yeah I think we should go the other way and add this to pretraining while keep backward compat for pretraining. Basically like this way of configuring it more.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**